### PR TITLE
Add subscribeallevents and unsubscribeallevents operations - closes #1082

### LINKED ISSUE #1082: Add `subscribeallevents` and `unsubscribeallevents` operations
In #133 we discussed how the current wot-discovery draft provides an [example Thing Description](https://w3c.github.io/wot-discovery/#directory-thing-description) for the Directory Service API which combines three different event types (`created_td`, `updated_td` and `deleted_td`) into a single `registration` event such that a consumer can subscribe and unsubscribe to all three events at once.

This feels like a workaround for a missing feature in the Thing Description specification which enables consumers to subscribe to all events or unsubscribe from all events in a single operation. I would therefore like to propose new `subscribeallevents` and `unsubscribeallevents` operations to fulfill this requirement.

I suggest these should be form operations which can be used on a top level form, similar to how `readallproperties` and `writeallproperties` currently works.

```json
{
    "@context": "http://www.w3.org/ns/td",
    "id": "urn:dev:ops:32473-WoTLamp-1234",
    "title": "MyLampThing",
    "securityDefinitions": {
        "basic_sc": {"scheme": "basic", "in": "header"}
    },
    "security": "basic_sc",
    "properties": {
        "on": {
            "type": "boolean",
            "forms": [{"href": "https://mylamp.example.com/properties/on"}]
        },
        "level": {
            "type": "number",
            "min": "0",
            "max": "100",
            "forms": [{"href": "https://mylamp.example.com/properties/on"}]
        }
    },
    "actions": {
        "toggle": {
            "forms": [{"href": "https://mylamp.example.com/actions/toggle"}]
        }
    },
    "events":{
        "overheated":{
            "forms": [{
                "href": "https://mylamp.example.com/events/overheated",
                "subprotocol": "longpoll"
            }]
        },
        "surge":{
            "forms": [{
                "href": "https://mylamp.example.com/events/surge",
                "subprotocol": "longpoll"
            }]
        }
    },
    "forms": [{
        "op": "readallproperties",
        "href": "https://mylamp.example.com/properties"
    }, {
        "op": "writeallproperties",
        "href": "https://mylamp.example.com/properties"
    },
    {
        "op": "subscribeallevents",
        "href": "https://mylamp.example.com/events",
        "subprotocol": "longpoll"
    },
    {
        "op": "unsubscribeallevents",
        "href": "https://mylamp.example.com/events",
        "subprotocol": "longpoll"
    }]
}
```

There would need to be some normative statements in the specification like there currently is for `readallproperties` and `writeallproperties` which describe what the payload for such resources should look like (i.e. a concatenation of the individual event interaction payloads, keyed by event name).

See also: This is related to the `readallpastevents` operation proposed in https://github.com/w3c/wot-thing-description/issues/892#issuecomment-802058379 for event logs.

### PR DESCRIPTION
**Pull request by benfrancis**
As discussed in #1082.

Note that these new ops are needed for the Directory Service API in the WoT Discovery specification (see https://github.com/w3c/wot-discovery/pull/159), and for making WebThings Gateway W3C compliant (see https://github.com/WebThingsIO/gateway/issues/2802#issuecomment-848736271).

Also fixed a typo on observeallproperties in the ontology.

***
<a href="https://pr-preview.s3.amazonaws.com/benfrancis/wot-thing-description/pull/1191.html" title="Last updated on Jul 14, 2021, 2:11 PM UTC (e254d2c)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/wot-thing-description/1191/8dda72f...benfrancis:e254d2c.html" title="Last updated on Jul 14, 2021, 2:11 PM UTC (e254d2c)">Diff</a>

### CODE REVIEW

**Review by relu91 — APPROVED** [review]
PR looks good. Two comments:
- The changes somehow touched also unexpected files (like visualizaton/tm.svg) and the preview does not show the TD model graph anymore. Is it a bug in the preview renderer? 
- The other operation types are described with a short sentence in [wot-architecture](https://www.w3.org/TR/wot-architecture/#sec-hypermedia-forms). In the very same document an editor's note states that `Other specifications may define further well-known operation types that are valid for their respective document format or form serialization` should we spend few words to explain `subscribeallevent`, `unsubscribeallevent` etc. in the TD document? It might be beneficial to have also examples about this new operations, like the one with SSE.

**Comment by benfrancis** [comment]
Thanks for the review @relu91.

> The changes somehow touched also unexpected files (like visualizaton/tm.svg) and the preview does not show the TD model graph anymore. Is it a bug in the preview renderer?

I found it quite hard to figure out which files are meant to be edited manually and which are meant to be generated by `render.sh`.

I used https://github.com/w3c/wot-thing-description/commit/5ac3b21e1431f859bc6af43fd2bc710bc629756c as a guide to figure out which files needed to change, then tried to figure out from the comments in the rendering script which files to edit vs. generate. The files I edited manually were:

1. `context/hypermedia-context.jsonld`
2. `validation/td-validation.ttl`
3. `validation/ext-td-json-schema-validation.json`
4. `validation/td-json-schema-validation.json`
3. `ontology/td.ttl`
4. `index.template.html`

All the other changes were automatically generated when running `npm run render`. Please advise whether I should do any of that differently or exclude any files from the diff.

> The other operation types are described with a short sentence in [wot-architecture](https://www.w3.org/TR/wot-architecture/#sec-hypermedia-forms). In the very same document an editor's note states that `Other specifications may define further well-known operation types that are valid for their respective document format or form serialization` should we spend few words to explain `subscribeallevent`, `unsubscribeallevent` etc. in the TD document? It might be beneficial to have also examples about this new operations, like the one with SSE.

I think we may have discussed this before but in my view Table 1 from [7.6.2 Forms](https://w3c.github.io/wot-architecture/#sec-hypermedia-forms) in the WoT Architecture specification should be moved to the Thing Description specification since it's not immediately obvious from the Thing Description specification where to find those definitions or even that they exist.

I'm happy to add a description of `subscribeallevents` and `unsubscribeallevents` somewhere in the Thing Description specification, but it would probably be better to keep them all in one place?

**Comment by relu91** [comment]
> Thanks for the review @relu91.
> 
> > The changes somehow touched also unexpected files (like visualizaton/tm.svg) and the preview does not show the TD model graph anymore. Is it a bug in the preview renderer?
> 
> I found it quite hard to figure out which files are meant to be edited manually and which are meant to be generated by `render.sh`.
> 
> I used [5ac3b21](https://github.com/w3c/wot-thing-description/commit/5ac3b21e1431f859bc6af43fd2bc710bc629756c) as a guide to figure out which files needed to change, then tried to figure out from the comments in the rendering script which files to edit vs. generate. The files I edited manually were:
> 
> 1. `context/hypermedia-context.jsonld`
> 2. `validation/td-validation.ttl`
> 3. `validation/ext-td-json-schema-validation.json`
> 4. `validation/td-json-schema-validation.json`
> 5. `ontology/td.ttl`
> 6. `index.template.html`
> 
> All the other changes were automatically generated when running `npm run render`. Please advise whether I should do any of that differently or exclude any files from the diff.

Yeah, I feel you. it would be better to have some guidance in a readme file, plus some live PR validation using a GitHub action. To me, the file list seems correct. Maybe some `npm install` artifacts? (i.e., it installed a newer version of the graph rendering tool which caused the changes?)

> > The other operation types are described with a short sentence in [wot-architecture](https://www.w3.org/TR/wot-architecture/#sec-hypermedia-forms). In the very same document an editor's note states that `Other specifications may define further well-known operation types that are valid for their respective document format or form serialization` should we spend few words to explain `subscribeallevent`, `unsubscribeallevent` etc. in the TD document? It might be beneficial to have also examples about this new operations, like the one with SSE.
> 
> I think we may have discussed this before but in my view Table 1 from [7.6.2 Forms](https://w3c.github.io/wot-architecture/#sec-hypermedia-forms) in the WoT Architecture specification should be moved to the Thing Description specification since it's not immediately obvious from the Thing Description specification where to find those definitions or even that they exist.

I remember that we mentioned this issue before, maybe in profile. The current direction seems to have most of the concepts in the architecture document and then refer back to it from the TD spec. I don't have a strong opinion about this as long as all the operations are correctly defined/explained 😄 . It might make sense to do this in another PR then. 

> I'm happy to add a description of `subscribeallevents` and `unsubscribeallevents` somewhere in the Thing Description specification, but it would probably be better to keep them all in one place?
> 

Yes I would prefer to have them all in one place 👍🏻 

> Edit: I think `observeallproperties` and `unobserveallproperties` are also currently missing from that table?

yes, they are missing too!

File: ontology/td.ttl [hunk 1]
```diff
@@ -239,7 +239,7 @@
     rdfs:comment """Operation type of forms used to read a property value"""@en ;
     rdfs:isDefinedBy : .
 
-:observeAllbserveProperties rdf:type :OperationType ;
+:observeAllProperties rdf:type :OperationType ;
     rdfs:label "observeAllbserveProperties" ;
     rdfs:comment """Operation type of forms used to observe all property values"""@en ;
```
Comment (danielpeintner) [inline ontology/td.ttl:244]: `rdfs:label` still has the typo

**Comment by benfrancis** [comment]
@relu91 wrote:
> Maybe some `npm install` artifacts? (i.e., it installed a newer version of the graph rendering tool which caused the changes?)

Could be, yes. I'm not sure how to fix that. Shall I remove some of the generated files from the diff?

> Yes I would prefer to have them all in one place 👍🏻

OK, let's create a follow-up PR for Architecture to add descriptions for the missing options, since that's the current home of the table.

**Comment by benfrancis** [comment]
@relu91 Which generated files do you think should be removed?

**Comment by benfrancis** [comment]
OK, I reverted the SVG changes.

**Review by sebastiankb — APPROVED** [review]

**Comment by sebastiankb** [comment]
from today's TD call, group decided to merge this PR

**Pull request merged**